constraint-solver/src/solver.rs

```diff
@@ -0,0 +1,210 @@
+use powdr_number::FieldElement;
+
+use crate::range_constraint::RangeConstraint;
+use crate::utils::known_variables;
+use crate::variable_update::VariableUpdate;
+
+use super::effect::Effect;
+use super::quadratic_symbolic_expression::{Error, RangeConstraintProvider};
+use super::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    symbolic_expression::SymbolicExpression,
+};
+use std::collections::BTreeMap;
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+/// The result of the solving process.
+#[allow(dead_code)]
+pub struct SolveResult<T: FieldElement, V> {
+    /// The concrete variable assignments that were derived.
+    pub assignments: BTreeMap<V, T>,
+    /// The final state of the algebraic constraints, with known variables
+    /// replaced by their values.
+    pub simplified_algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>,
+}
+
+/// Given a list of constraints, tries to derive as many variable assignments as possible.
+pub struct Solver<T: FieldElement, V> {
+    /// The algebraic constraints to solve.
+    /// Note that these are mutated during the solving process.
+    /// They must be kept consistent with the variable states.
+    algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>,
+    /// The current state of the variables.
+    variable_states: BTreeMap<V, VariableState<T>>,
+}
+
+impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Solver<T, V> {
+    #[allow(dead_code)]
+    pub fn new(algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>) -> Self {
+        assert!(
+            known_variables(&algebraic_constraints).is_empty(),
+            "Expected all variables to be unknown."
+        );
+
+        Solver {
+            algebraic_constraints,
+            variable_states: BTreeMap::new(),
+        }
+    }
+
+    /// Solves the constraints as far as possible, returning concrete variable
+    /// assignments and a simplified version of the algebraic constraints.
+    #[allow(dead_code)]
+    pub fn solve(mut self) -> Result<SolveResult<T, V>, Error> {
+        self.loop_until_no_progress()?;
+
+        let assignments = self
+            .variable_states
+            .into_iter()
+            .filter_map(|(v, info)| match info {
+                VariableState::Known(expr) => Some((v, expr)),
+                VariableState::Unknown(_) => None,
+            })
+            .collect();
+        Ok(SolveResult {
+            assignments,
+            simplified_algebraic_constraints: self.algebraic_constraints,
+        })
+    }
+
+    fn loop_until_no_progress(&mut self) -> Result<(), Error> {
+        loop {
+            let mut progress = false;
+            for i in 0..self.algebraic_constraints.len() {
```

Comment:
Do you loop over the indexes because of the borrow checker?

Reply:
Yes. Can't be borrowing anything from `self` when calling `apply_effects`.